### PR TITLE
Resolves Issue 110

### PR DESCRIPTION
**Pull request by arthi-j**
*Issue #, if available:* https://github.com/aws-samples/serverless-test-samples/issues/110

*Description of changes:* Python sample code for schema and contract testing

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: python-test-samples/schema-and-contract-testing/tests/test_schema.py
Comment (wtromano-aws) [inline python-test-samples/schema-and-contract-testing/tests/test_schema.py:1]: Pylint:

test_schema.py:27:0: C0301: Line too long (107/100) (line-too-long)
test_schema.py:53:0: C0301: Line too long (126/100) (line-too-long)
test_schema.py:1:0: C0114: Missing module docstring (missing-module-docstring)
test_schema.py:5:0: E0401: Unable to import 'pytest' (import-error)
test_schema.py:7:0: E0401: Unable to import 'pydantic' (import-error)
test_schema.py:15:0: C0116: Missing function or method docstring (missing-function-docstring)
test_schema.py:17:9: W1514: Using open without explicitly specifying an encoding (unspecified-encoding)
test_schema.py:17:41: C0103: Variable name "f" doesn't conform to snake_case naming style (invalid-name)
test_schema.py:24:0: C0116: Missing function or method docstring (missing-function-docstring)
test_schema.py:31:0: C0116: Missing function or method docstring (missing-function-docstring)
test_schema.py:38:0: C0116: Missing function or method docstring (missing-function-docstring)
test_schema.py:40:8: W0612: Unused variable 'customer' (unused-variable)
test_schema.py:51:0: C0116: Missing function or method docstring (missing-function-docstring)
test_schema.py:7:0: C0411: third party import "from pydantic import ValidationError" should be placed before "from tests.context import CustomerModel" (wrong-import-order)

File: python-test-samples/schema-and-contract-testing/tests/test_contract.py
Comment (wtromano-aws) [inline python-test-samples/schema-and-contract-testing/tests/test_contract.py:1]: Pylint:

test_contract.py:19:0: C0301: Line too long (107/100) (line-too-long)
test_contract.py:1:0: C0114: Missing module docstring (missing-module-docstring)
test_contract.py:2:0: E0401: Unable to import 'pytest' (import-error)
test_contract.py:4:0: E0401: Unable to import 'pydantic' (import-error)
test_contract.py:8:0: C0116: Missing function or method docstring (missing-function-docstring)
test_contract.py:10:9: W1514: Using open without explicitly specifying an encoding (unspecified-encoding)
test_contract.py:10:41: C0103: Variable name "f" doesn't conform to snake_case naming style (invalid-name)
test_contract.py:17:0: C0116: Missing function or method docstring (missing-function-docstring)
test_contract.py:24:0: C0116: Missing function or method docstring (missing-function-docstring)
test_contract.py:26:8: W0612: Unused variable 'customer' (unused-variable)
test_contract.py:4:0: C0411: third party import "from pydantic import ValidationError" should be placed before "from tests.context import ConsumerCustomerModel" (wrong-import-order)

File: python-test-samples/schema-and-contract-testing/tests/context.py
Comment (wtromano-aws) [inline python-test-samples/schema-and-contract-testing/tests/context.py:1]: Pylint:

context.py:1:0: C0114: Missing module docstring (missing-module-docstring)
context.py:16:0: E0401: Unable to import 'customer_v1_0_0_with_validation' (import-error)
context.py:16:0: C0413: Import "from customer_v1_0_0_with_validation import Customer as ConsumerCustomerModel" should be placed at the top of the module (wrong-import-position)
context.py:17:0: E0401: Unable to import 'customer_v1_0_0' (import-error)
context.py:17:0: C0413: Import "from customer_v1_0_0 import Customer as CustomerModel" should be placed at the top of the module (wrong-import-position)
context.py:16:0: W0611: Unused Customer imported from customer_v1_0_0_with_validation as ConsumerCustomerModel (unused-import)

File: python-test-samples/schema-and-contract-testing/src/customer_v1_0_0_with_validation.py
Comment (wtromano-aws) [inline python-test-samples/schema-and-contract-testing/src/customer_v1_0_0_with_validation.py:1]: Pylint:

customer_v1_0_0.py:1:0: C0114: Missing module docstring (missing-module-docstring)
customer_v1_0_0.py:9:0: C0115: Missing class docstring (missing-class-docstring)
customer_v1_0_0.py:9:0: R0903: Too few public methods (0/2) (too-few-public-methods)

File: python-test-samples/schema-and-contract-testing/src/customer_v1_1_0.py
Comment (wtromano-aws) [inline python-test-samples/schema-and-contract-testing/src/customer_v1_1_0.py:1]: Pylint:

customer_v1_2_0.py:1:0: C0114: Missing module docstring (missing-module-docstring)
customer_v1_2_0.py:8:0: E0401: Unable to import 'pydantic' (import-error)
customer_v1_2_0.py:11:0: C0115: Missing class docstring (missing-class-docstring)
customer_v1_2_0.py:11:0: R0903: Too few public methods (0/2) (too-few-public-methods)

File: python-test-samples/schema-and-contract-testing/src/customer_v1_2_0.py
Comment (wtromano-aws) [inline python-test-samples/schema-and-contract-testing/src/customer_v1_2_0.py:1]: customer_v1_2_0.py:1:0: C0114: Missing module docstring (missing-module-docstring)
customer_v1_2_0.py:8:0: E0401: Unable to import 'pydantic' (import-error)
customer_v1_2_0.py:11:0: C0115: Missing class docstring (missing-class-docstring)
customer_v1_2_0.py:11:0: R0903: Too few public methods (0/2) (too-few-public-methods)

File: python-test-samples/schema-and-contract-testing/src/customer_v1_3_0.py
Comment (wtromano-aws) [inline python-test-samples/schema-and-contract-testing/src/customer_v1_3_0.py:1]: Pylint:

customer_v1_3_0.py:1:0: C0114: Missing module docstring (missing-module-docstring)
customer_v1_3_0.py:8:0: E0401: Unable to import 'pydantic' (import-error)
customer_v1_3_0.py:11:0: C0115: Missing class docstring (missing-class-docstring)
customer_v1_3_0.py:11:0: R0903: Too few public methods (0/2) (too-few-public-methods)

File: python-test-samples/schema-and-contract-testing/src/customer_v1_4_0.py
Comment (wtromano-aws) [inline python-test-samples/schema-and-contract-testing/src/customer_v1_4_0.py:1]: Pylint (Although the last error is probably intentional and can be ignored)

customer_v1_4_0.py:1:0: C0114: Missing module docstring (missing-module-docstring)
customer_v1_4_0.py:8:0: E0401: Unable to import 'pydantic' (import-error)
customer_v1_4_0.py:11:0: C0115: Missing class docstring (missing-class-docstring)
customer_v1_4_0.py:11:0: R0903: Too few public methods (0/2) (too-few-public-methods)
customer_v1_4_0.py:1:0: R0801: Similar lines in 2 files
==src.customer_v1_0_0:[8:17]
==src.customer_v1_0_0_with_validation:[8:20]

**Review by wtromano-aws — CHANGES_REQUESTED** [review]
Great pattern, thank you so much for contributing!  It looks great as-is, but I have some minor formatting and cosmetic improvement comments that are very minor in scope.  Certainly let me know of any questions on the feedback.

File: python-test-samples/schema-and-contract-testing/README.md
[the remark this answers is not in the document]
Comment (arthi-j) [reply python-test-samples/schema-and-contract-testing/README.md:1]: Done

File: python-test-samples/schema-and-contract-testing/README.md [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+### System Under Test (SUT)
+
+This test is integration service agnostic, as it runs against the schema published to a schema registry. Amazon EventBridge Schema Registry is one example of such a registry. It provides schemas for all events that are generated by AWS services. When you turn on schema discovery it will automatically detect and provide schemas for your custom events (OpenAPI and JSON Schemas). You can also use your own Git repository as a registry for storing schemas. In our test we will use a local “schemas” folder as a registry for simplicity.
+
+The diagram below shows an example of how schema tests may run as part of a producer’s CI/CD pipeline. You can [view a sample schema test](tests/test_schema.py) that could be run during the "Build" step of the process.
+
+![System Under Test (SUT)](img/schema_testing.png "Schema testing")
+
```
[the remark this answers is not in the document]
Comment (arthi-j) [reply python-test-samples/schema-and-contract-testing/README.md:49]: Done

File: python-test-samples/schema-and-contract-testing/README.md [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+---
+
+## Contents
+
+- [Python: Schema and Contract Testing for Event-Driven Architectures](#python-schema-and-contract-testing-for-event-driven-architectures)
+  - [Contents](#contents)
+  - [Introduction](#introduction)
+  - [Schema Testing](#schema-testing)
+    - [Goal](#goal)
```
[the remark this answers is not in the document]
Comment (arthi-j) [reply python-test-samples/schema-and-contract-testing/README.md:14]: Done

File: python-test-samples/schema-and-contract-testing/README.md [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+To prevent event schema changes from breaking consumer applications.
+
+### Description
+
+Schema tests are designed to check for changes that are not backward-compatible, such as removing elements, changing a field type or adding new mandatory elements. Adding optional elements is considered backward-compatible. If a new schema version is backward-compatible, the schema tests will pass and the consumer application business logic should not fail.
+
+### System Under Test (SUT)
+
+This test is integration service agnostic, as it runs against the schema published to a schema registry. Amazon EventBridge Schema Registry is one example of such a registry. It provides schemas for all events that are generated by AWS services. When you turn on schema discovery it will automatically detect and provide schemas for your custom events (OpenAPI and JSON Schemas). You can also use your own Git repository as a registry for storing schemas. In our test we will use a local “schemas” folder as a registry for simplicity.
```
[the remark this answers is not in the document]
Comment (arthi-j) [reply python-test-samples/schema-and-contract-testing/README.md]: Done

File: python-test-samples/schema-and-contract-testing/tests/test_schema.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+import json
+import pytest
+from tests.context import CustomerModel
+from pydantic import ValidationError
+
+# v1.0.0 represents the current schema in use and is the baseline to detect breaking changes
+# v1.1.0 to v1.4.0 represent changes from v1.0.0
+
+# utility method to instantiate customer model 1.0.0 with event passed as
+# parameter
```
[the remark this answers is not in the document]
Comment (arthi-j) [reply python-test-samples/schema-and-contract-testing/tests/test_schema.py]: Done

File: python-test-samples/schema-and-contract-testing/src/customer_v1_4_0.py
Comment (arthi-j) [reply python-test-samples/schema-and-contract-testing/src/customer_v1_4_0.py:1]: Not all Pylint recommendations make sense. Updated where relevant.

**Review by wtromano-aws — APPROVED** [review]

**Pull request merged**